File: ios/StelliveHubiOS/StelliveHubiOS/Models/ReservationModels.swift
import Foundation

enum ReservationKind: String, Codable, CaseIterable, Hashable {
    case ticket
    case purchase
    case reservation

    var displayName: String {
        switch self {
        case .ticket: "티켓"
        case .purchase: "구매"
        case .reservation: "예약"
        }
    }
}

enum ReservationStatus: String, Codable, CaseIterable, Hashable {
    case pendingConfirmation
    case confirmed
    case cancelled
    case refunded
    case completed

    var displayName: String {
        switch self {
        case .pendingConfirmation: "확인 필요"
        case .confirmed: "예약 완료"
        case .cancelled: "취소"
        case .refunded: "환불"
        case .completed: "이용 완료"
        }
    }
}

enum ReservationLinkSource: String, Codable, Hashable {
    case appInput
    case browserShare
    case liveActivity
    case control
}

struct ReservationEventSnapshot: Codable, Hashable {
    let title: String
    let category: String
    let startsAt: Date?
    let endsAt: Date?
    let venueName: String?
    let venueAddress: String?
    let sourceLabel: String
    let imageURL: String?
}

struct ReservationDraft: Codable, Identifiable, Hashable {
    var id: UUID { sessionID }
    let sessionID: UUID
    let eventID: String
    let scheduleItemID: String?
    let kind: ReservationKind
    let eventSnapshot: ReservationEventSnapshot
    let originalActionURL: String
    let providerHost: String
    var openedAt: Date
    var expiresAt: Date
    var attemptCount: Int
}

struct ReservationRecord: Codable, Identifiable, Hashable {
    let id: UUID
    let sourceSessionID: UUID?
    let eventID: String?
    let scheduleItemID: String?
    var kind: ReservationKind
    var status: ReservationStatus
    let eventSnapshot: ReservationEventSnapshot
    let originalActionURL: String?
    var reservationDetailURL: String?
    var providerHistoryURL: String?
    var linkSource: ReservationLinkSource?
    var displayTitleOverride: String?
    var startsAtOverride: Date?
    var endsAtOverride: Date?
    var venueOverride: String?
    var optionText: String?
    var quantity: Int?
    var referenceNumber: String?
    var note: String?
    let openedAt: Date?
    let confirmedAt: Date
    let createdAt: Date
    var updatedAt: Date
    let schemaVersion: Int

    var displayTitle: String {
        ReservationTextPolicy.nonEmpty(displayTitleOverride) ?? eventSnapshot.title
    }

    var effectiveStartsAt: Date? { startsAtOverride ?? eventSnapshot.startsAt }
    var effectiveEndsAt: Date? { endsAtOverride ?? eventSnapshot.endsAt }
    var effectiveVenue: String? { ReservationTextPolicy.nonEmpty(venueOverride) ?? eventSnapshot.venueName }

    var preferredURL: URL? {
        [reservationDetailURL, providerHistoryURL, originalActionURL]
            .compactMap { ReservationURLPolicy.validatedURL($0) }
            .first
    }
}

enum ReservationTextPolicy {
    static func nonEmpty(_ value: String?) -> String? {
        guard let value else { return nil }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

enum ReservationURLPolicy {
    static let maximumLength = 4_096
    private static let sensitiveKeys = Set(["token", "auth", "session", "signature", "code"])

    static func validatedURL(_ rawValue: String?) -> URL? {
        guard let value = ReservationTextPolicy.nonEmpty(rawValue), value.count <= maximumLength,
              let components = URLComponents(string: value),
              components.scheme?.lowercased() == "https",
              components.host?.isEmpty == false,
              let url = components.url
        else { return nil }
        return url
    }

    static func containsSensitiveQuery(_ rawValue: String?) -> Bool {
        guard let url = validatedURL(rawValue),
              let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems
        else { return false }
        return items.contains { sensitiveKeys.contains($0.name.lowercased()) }
    }

    static func firstSharedHTTPSURL(in text: String?) -> URL? {
        guard let text else { return nil }
        let detector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)
        let range = NSRange(text.startIndex..., in: text)
        return detector?.matches(in: text, range: range)
            .compactMap(\.url)
            .first(where: { validatedURL($0.absoluteString) != nil })
    }
}

enum ReservationDraftSelection: Equatable {
    case none
    case selected(UUID)
    case choose([UUID])
}

enum ReservationDraftPolicy {
    static func active(_ drafts: [ReservationDraft], now: Date = Date()) -> [ReservationDraft] {
        drafts.filter { $0.expiresAt > now }.sorted { $0.openedAt > $1.openedAt }
    }

    static func selection(from drafts: [ReservationDraft], requested: UUID? = nil, now: Date = Date()) -> ReservationDraftSelection {
        let activeDrafts = active(drafts, now: now)
        if let requested, activeDrafts.contains(where: { $0.sessionID == requested }) { return .selected(requested) }
        switch activeDrafts.count {
        case 0: return .none
        case 1: return .selected(activeDrafts[0].sessionID)
        default: return .choose(activeDrafts.map(\.sessionID))
        }
    }
}

File: android/StelliveHubAndroid/app/src/main/java/dev/minepacu/stelliveeventnotifier/feature/reservations/domain/ReservationModels.kt
package dev.minepacu.stelliveeventnotifier.feature.reservations.domain

import java.time.Instant
import java.util.UUID

enum class ReservationKind { TICKET, PURCHASE, RESERVATION }

enum class ReservationStatus { PENDING_CONFIRMATION, CONFIRMED, CANCELLED, REFUNDED, COMPLETED }

enum class ReservationLinkSource { APP_INPUT, BROWSER_SHARE, SYSTEM_SHORTCUT }

data class ReservationEventSnapshot(
    val title: String,
    val category: String,
    val startsAt: Instant? = null,
    val endsAt: Instant? = null,
    val venueName: String? = null,
    val venueAddress: String? = null,
    val sourceLabel: String,
    val imageUrl: String? = null,
)

data class ReservationDraft(
    val sessionId: UUID,
    val eventId: String,
    val scheduleItemId: String?,
    val kind: ReservationKind,
    val eventSnapshot: ReservationEventSnapshot,
    val originalActionUrl: String,
    val providerHost: String,
    val openedAt: Instant,
    val expiresAt: Instant,
    val attemptCount: Int,
)

data class ReservationRecord(
    val id: UUID,
    val sourceSessionId: UUID?,
    val eventId: String?,
    val scheduleItemId: String?,
    val kind: ReservationKind,
    val status: ReservationStatus,
    val eventSnapshot: ReservationEventSnapshot,
    val originalActionUrl: String?,
    val reservationDetailUrl: String?,
    val providerHistoryUrl: String?,
    val linkSource: ReservationLinkSource?,
    val displayTitleOverride: String?,
    val startsAtOverride: Instant?,
    val endsAtOverride: Instant?,
    val venueOverride: String?,
    val optionText: String?,
    val quantity: Int?,
    val referenceNumber: String?,
    val note: String?,
    val openedAt: Instant?,
    val confirmedAt: Instant,
    val createdAt: Instant,
    val updatedAt: Instant,
    val schemaVersion: Int = 1,
) {
    val displayTitle: String get() = displayTitleOverride?.takeIf(String::isNotBlank) ?: eventSnapshot.title
    val effectiveStartsAt: Instant? get() = startsAtOverride ?: eventSnapshot.startsAt
    val effectiveEndsAt: Instant? get() = endsAtOverride ?: eventSnapshot.endsAt
    val effectiveVenue: String? get() = venueOverride?.takeIf(String::isNotBlank) ?: eventSnapshot.venueName
    val preferredOpenUrl: String? get() = reservationDetailUrl ?: providerHistoryUrl ?: originalActionUrl
}
